src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/Rfc2898DeriveBytes.OneShot.cs

```diff
@@ -0,0 +1,306 @@
+        ///   <paramref name="hashAlgorithm" /> has a <see cref="HashAlgorithmName.Name" />
+        ///   that is empty or <see langword="null" />.
+        /// </exception>
+        /// <exception cref="CryptographicException">
+        ///   <paramref name="hashAlgorithm" /> is an unsupported hash algorithm. Supported algorithms
+        ///   are <see cref="HashAlgorithmName.SHA1" />, <see cref="HashAlgorithmName.SHA256" />,
+        ///   <see cref="HashAlgorithmName.SHA384" />, and <see cref="HashAlgorithmName.SHA512" />.
+        /// </exception>
+        /// <exception cref="EncoderFallbackException">
+        /// <paramref name="password" /> contains text that cannot be converted to UTF8.
+        /// </exception>
+        /// <remarks>
+        /// The <paramref name="password" /> will be converted to bytes using the UTF8 encoding. For
+        /// other encodings, convert the password string to bytes using the appropriate <see cref="System.Text.Encoding" />
+        /// and use <see cref="Pbkdf2(ReadOnlySpan{byte}, ReadOnlySpan{byte}, Span{byte}, int, HashAlgorithmName)" />.
+        /// </remarks>
+        public static void Pbkdf2(
+            ReadOnlySpan<char> password,
+            ReadOnlySpan<byte> salt,
+            Span<byte> destination,
+            int iterations,
+            HashAlgorithmName hashAlgorithm)
+        {
+            const int MaxPasswordStackSize = 256;
+
+            byte[]? rentedPasswordBuffer = null;
+            int maxEncodedSize = s_throwingUtf8Encoding.GetMaxByteCount(password.Length);
+
+            Span<byte> passwordBuffer = maxEncodedSize > MaxPasswordStackSize ?
+                (rentedPasswordBuffer = CryptoPool.Rent(maxEncodedSize)) :
+                stackalloc byte[MaxPasswordStackSize];
+            int passwordBytesWritten = s_throwingUtf8Encoding.GetBytes(password, passwordBuffer);
+            Span<byte> passwordBytes = passwordBuffer.Slice(0, passwordBytesWritten);
+
+            try
+            {
+                Pbkdf2Core(passwordBytes, salt, destination, iterations, hashAlgorithm);
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(passwordBytes);
+            }
+
+            if (rentedPasswordBuffer is not null)
+            {
+                CryptoPool.Return(rentedPasswordBuffer, clearSize: 0); // manually cleared above.
+            }
+        }
+
+        private static void Pbkdf2Core(
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            Span<byte> destination,
+            int iterations,
+            HashAlgorithmName hashAlgorithm)
+        {
+            if (iterations <= 0)
+                throw new ArgumentOutOfRangeException(nameof(iterations), SR.ArgumentOutOfRange_NeedPosNum);
+            if (string.IsNullOrEmpty(hashAlgorithm.Name))
+                throw new ArgumentException(SR.Cryptography_HashAlgorithmNameNullOrEmpty, nameof(hashAlgorithm));
+
```

Comment:
Ideally this would be done before any of the array-allocating versions allocate their array.  Just so that the OOM for asking for too large of an output on a constrained system doesn't happen before noticing that there were invalid inputs.  Then this method should just assert (typically true of a `*Core` method anyways)

Reply:
@bartonjs ah, yeah, makes sense. Refactored things a bit to make that work.